### DISCUSSION TITLE
OrderBy Average Field

### DISCUSSION THREAD

**Discussion by un-focused** [prisma/prisma#6361] [Q&A] [answered]
If I have a User model that contains reviews, is it possible to write a prisma style query to get users sorted by the average rating descending so that I get the most highly rated users first? In SQL I would use the AVG function and then use GroupBy and OrderBy but is there an equivalent in Prisma??

Here is a basic version of the models that I am using:
```
model User {
    id       Int                    @id @default(autoincrement())
    username String                 @unique
    // reviews that this person has rated (on other users' profile's)
    reviewed UserProfileReview[]    @relation(name: "Reviewer")
    // reviews where this person has been reviewed (on this users' profile)
    reviews  UserProfileReview[]    @relation(name: "Reviewee")
}

model UserProfileReview {
    id          Int                        @id @default(autoincrement())
    rating      Int                        @db.SmallInt
    description String?
    reviewerID  Int                        @default(autoincrement())
    revieweeID  Int                        @default(autoincrement())
    reviewer    UserProfile                @relation(name: "Reviewer", fields: [reviewerID], references: [id])
    reviewee    UserProfile                @relation(name: "Reviewee", fields: [revieweeID], references: [id])

    @@unique([reviewerID, revieweeID])
}
```

This is what I have so far:
```
this.prisma.userProfile.findMany(
    {
        select: {
            username: true,
            reviews: true,
        },
        orderBy: {},
        take: 9
    }
);
```
Not exactly sure what to put in OrderBy so that I can access the fields of reviews or if that is even possible in one query in Prisma.
I've been looking at the following pages:
[Relations](https://www.prisma.io/docs/concepts/components/prisma-client/relation-queries)
[Aggregate](https://www.prisma.io/docs/reference/api-reference/prisma-client-reference#aggregate)
[Summation](https://www.prisma.io/docs/concepts/components/prisma-client/aggregation-grouping-summarizing#count)

If someone could point me in the right direction or tell me if it is possible with Prisma that would be much appreciated! If it is not supported, I will raise an issue!

**Comment by 2color** [comment, marked as answer]
Hey @un-focused 👋,

The following query will give you the average rating per reviewee:

```ts
const users = await prisma.userProfileReview.groupBy({
    by: ['revieweeID'],
    avg: {
      rating: true,
    },
  })
```

This will return the following:
```
[
  { revieweeID: 4, avg: { rating: 10 } },
  { revieweeID: 2, avg: { rating: 7 } },
  { revieweeID: 1, avg: { rating: 8 } }
]
````


However, currently, it's still not possible to sort the results by the aggregate. We have an [open issue](https://github.com/prisma/prisma/issues/5464) for this which you can follow so you'll be notified once this feature is available.

**Reply by un-focused** [reply]
Thanks @2color ! Yeah that's what I figured, I was hoping for a way to combine them. I'll take a look at the issue 😄

**Reply by 2color** [reply]
Hey @un-focused 👋,

We just released version 2.21.0 of Prisma which adds Preview support for ordering by an aggregate in groupBy queries.

This should make it possible for you to order by the average.

**Reply by un-focused** [reply]
This is amazing @2color !! I've used libraries in the past and love open source. Usually the libraries I have worked with end up deprecated to not maintained well so I code manually but this is brilliant. If there are any tasks that you would want me to work on, let me know!

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
